zookeeper-api/src/main/java/org/apache/helix/zookeeper/impl/client/ZkClient.java

```diff
@@ -85,7 +88,20 @@ public class ZkClient extends org.apache.helix.zookeeper.zkclient.ZkClient imple
    *            The JMX bean name will be: HelixZkClient.monitorType.monitorKey.monitorInstanceName.
    * @param monitorRootPathOnly
    *            Should only stat of access to root path be reported to JMX bean or path-specific stat be reported too.
+   * @param connectOnInit true if connect to ZK during initialization, otherwise user will need to call connect
+   *                      explicitly before talking to ZK.
    */
+  public ZkClient(IZkConnection zkConnection, Watcher watcher, int connectionTimeout, long operationRetryTimeout,
+      PathBasedZkSerializer zkSerializer,
+      String monitorType, String monitorKey, String monitorInstanceName,
+      boolean monitorRootPathOnly, boolean connectOnInit) {
+    super(zkConnection, watcher, operationRetryTimeout, zkSerializer, monitorType, monitorKey, monitorInstanceName,
+        monitorRootPathOnly);
+    if (connectOnInit) {
+      connect(connectionTimeout, watcher);
+    }
```

Comment:
I would suggest get these logic another different PR. One PR just for pure refactoring not changing logic / adding things.

The other one is for new features. This looks like a new feature instead of refactoring.

Reply:
This PR alone is fully backward compatible. The new feature is backward compatible, and will be used in meta-client implementation. 
(however, with our offline discussion, most of this PR will not be used with the change of implementation)